cmd/go-sbot/config.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/komkom/toml"
+	"go.mindeco.de/log/level"
+	loglib "log"
+	"os"
+	"strconv"
+)
+
+type SbotConfig struct {
+	ShsCap string `json:"shscap,omitempty"`
+	Hmac   string `json:"hmac,omitempty"`
+	Hops   uint   `json:"hops,omitempty"`
+
+	Repo     string `json:"repo,omitempty"`
+	DebugDir string `json:"debugdir,omitempty"`
+
+	MuxRPCAddress    string `json:"lis,omitempty"`
+	WebsocketAddress string `json:"wslis,omitempty"`
+	MetricsAddress   string `json:"debuglis,omitempty"`
+
+	NoUnixSocket       bool `json:"nounixsock"`
+	EnableAdvertiseUDP bool `json:"localdv"`
+	EnableDiscoveryUDP bool `json:"localdiscov"`
+	EnableEBT          bool `json:"enable-ebt"`
+	EnableFirewall     bool `json:"promisc"`
+	RepairFSBeforeStart bool `json:"repair"`
+
+	presence map[string]interface{}
+}
+
+func (config SbotConfig) Has(flagname string) bool {
+	_, ok := config.presence[flagname]
+	return ok
+}
+
+func ReadConfig(configPath string) SbotConfig {
+	var conf SbotConfig
+	conf.presence = make(map[string]interface{})
+
+	data, err := os.ReadFile(configPath)
+	var msg string
+	if err != nil {
+		msg = fmt.Sprintf("no config detected at %s; running without it", configPath)
+		level.Info(log).Log("event", msg)
```

Comment:
please don't use printf with the logfmt library but pure `"label", value` pairs. 

There are parsers for logfmt as well as other outputs like JSON so you don't have to comb through the `%s`s and `%v`'s by hand.

Reply:
:000

Reply:
oooookay! i think this is (way, _way_) more in the vein with structured logging approach (note: actual user has been replaced with `<usr>` below)

```
level=info t=28.912µs event="read config" msg="no config detected" path=/home/<usr>/.ssb-go/config.toml
level=info t=300.477µs event="write running-config.json" msg="active config and env vars have been persisted" path=/home/<usr>/.ssb-go/running-config.json
level=info t=323.635µs event="set repo" path=/home/<usr>/.ssb-go
```

honestly had no idea why you were going through the hoops of using the logger before, but now i can see a bit more why and how it could be useful :~)